kombu/common.py

```diff
@@ -105,6 +105,12 @@ def declaration_cached(entity, channel):
 
 def maybe_declare(entity, channel=None, retry=False, **retry_policy):
     """Declare entity (cached)."""
+    if retry:
+        return _imaybe_declare(entity, channel, **retry_policy)
+    return _maybe_declare(entity, channel)
+
+
+def _maybe_declare(entity, channel):
     is_bound = entity.is_bound
```

Comment:
Why was this code moved to _maybe_declare?

Reply:
@tyarimi thank you very much for analyzing the flow in detail. I have made the following observations and adding them here mostly for context:

- I think your explanation is correct, with the previous calling sequence, the `channel` argument value would remain the same. I am not sure however when the error will be raised here. thus triggering the retry loop. Perhaps it is the [initial check](https://github.com/celery/kombu/blob/master/kombu/common.py#L134) for non-`None` connection or during the call to `entity.declare` (following lines)?
- [entity.revive](https://github.com/celery/kombu/blob/master/kombu/abstract.py#L83) originates from `MaybeChannelBound` mixin, a class used in the superclass list of `entity.Exchange` and `entity.Queue`. 
- [messaging.Producer.maybe_declare](https://github.com/celery/kombu/blob/master/kombu/messaging.py#L102) uses a non-`None` value for the `channel` argument. `retry` is by default `False` for this proxy function and is called with default argument values in `_publish` method. I have not found other usages of `maybe_declare` in Kombu (or Celery).

The changes seem to be in the right direction and this could be a serious bug, I think we may have encountered it in other bug reports. @tyarimi if you can, please add log output that displays the stack traces during continuous reconnection attempts, as these could be very useful in pointing us on how exactly the flow is executed. Thanks.